[FINAL] [PICK ONE] Ike Particles

aab29 opened the pull request:
Accompanies https://github.com/mindsnacks/Ahu/pull/1490 .

Completely new particle system that actually works.  Based closely on the Cocos2D implementation.  Let me know if there's anything you'd like to switch over to our terminology.  :speech_balloon: 

Note that this PR completely removes the old system (```MOAIParticleSystem``` and accompanying classes).  This is done deliberately to avoid confusion.  :bomb:

aab29 commented on src/moaicore/MOAICCParticleSystem.cpp:
```diff
@@ -0,0 +1,2394 @@
+				
+				if ( this->mParticleCount == 0 ){
+					
+				}
```
TODO: Figure out what's up with this.  :zero:

aab29 commented:
I've been testing this out a bunch.  I'll mark it as **[FINAL]** as well.  :microscope:

kstenerud commented on src/moaicore/MOAICCParticleSystem.cpp:
```diff
@@ -0,0 +1,2389 @@
+ */
+int MOAICCParticleSystem::_getFinishParticleSize ( lua_State* L ) {
+	MOAI_LUA_SETUP ( MOAICCParticleSystem, "U" )
+	lua_pushnumber ( state, self->mFinishSize );
```
This should have the same name as the getter/setter.

kstenerud commented on src/moaicore/MOAICCParticleSystem.cpp:
```diff
@@ -0,0 +1,2389 @@
+ */
+int MOAICCParticleSystem::_getFrequency ( lua_State* L ) {
+	MOAI_LUA_SETUP ( MOAICCParticleSystem, "U" )
+	lua_pushnumber ( state, self->mEmissionRate );
```
Should settle on either frequency or emission rate.

kstenerud commented on src/moaicore/MOAICCParticleSystem.cpp:
```diff
@@ -0,0 +1,2389 @@
+ */
+int MOAICCParticleSystem::_getLifespan ( lua_State* L ) {
+	MOAI_LUA_SETUP ( MOAICCParticleSystem, "U" )
+	lua_pushnumber ( state, self->mLifespanTerm[0] );
```
lifespan vs lifespanTerm

kstenerud commented on src/moaicore/MOAICCParticleSystem.cpp:
```diff
@@ -0,0 +1,2389 @@
+ */
+int MOAICCParticleSystem::_setStartRadius ( lua_State *L ) {
+	MOAI_LUA_SETUP ( MOAICCParticleSystem, "UN" )
+	self->mMaxRadius = state.GetValue < float >( 2, 0 );
```
startRadius vs maxRadius

kstenerud commented on src/moaicore/MOAICCParticleSystem.cpp:
```diff
@@ -0,0 +1,2389 @@
+ */
+int MOAICCParticleSystem::_getStartRadiusVariance ( lua_State *L ) {
+	MOAI_LUA_SETUP ( MOAICCParticleSystem, "U" )
+	lua_pushnumber ( state, self->mMaxRadiusVariance );
```
startRadiusVariance vs maxRadiusVariance

kstenerud commented on src/moaicore/MOAICCParticleSystem.cpp:
```diff
@@ -0,0 +1,2389 @@
+ */
+int MOAICCParticleSystem::_getEndRadius ( lua_State *L ) {
+	MOAI_LUA_SETUP ( MOAICCParticleSystem, "U" )
+	lua_pushnumber ( state, self->mMinRadius );
```
minRadius vs endRadius

kstenerud commented on src/moaicore/MOAICCParticleSystem.cpp:
```diff
@@ -0,0 +1,2389 @@
+ */
+int MOAICCParticleSystem::_getEndRadiusVariance ( lua_State *L ) {
+	MOAI_LUA_SETUP ( MOAICCParticleSystem, "U" )
+	lua_pushnumber ( state, self->mMinRadiusVariance );
```
minRadiusVariance vs endRadiusVariance

kstenerud commented on src/moaicore/MOAICCParticleSystem.cpp:
```diff
@@ -0,0 +1,2389 @@
+ */
+int MOAICCParticleSystem::_getParticleLifespan ( lua_State *L ) {
+	MOAI_LUA_SETUP ( MOAICCParticleSystem, "U" )
+	lua_pushnumber ( state, self->mLifespan );
```
lifespan vs particleLifespan

kstenerud commented on src/moaicore/MOAICCParticleSystem.cpp:
```diff
@@ -0,0 +1,2389 @@
+ */
+int MOAICCParticleSystem::_getParticleLifespanVariance ( lua_State *L ) {
+	MOAI_LUA_SETUP ( MOAICCParticleSystem, "U" )
+	lua_pushnumber ( state, self->mLifespanVariance );
```
lifespanVariance vs particleLifespanVariance

kstenerud commented on src/moaicore/MOAICCParticleSystem.cpp:
```diff
@@ -0,0 +1,2389 @@
+ */
+int MOAICCParticleSystem::_getRotatePerSecond ( lua_State *L ) {
+	MOAI_LUA_SETUP ( MOAICCParticleSystem, "U" )
+	lua_pushnumber ( state, self->mRotPerSecond );
```
rotatePerSecond vs rotPerSecond

kstenerud commented on src/moaicore/MOAICCParticleSystem.cpp:
```diff
@@ -0,0 +1,2389 @@
+ */
+int MOAICCParticleSystem::_getRotatePerSecondVariance ( lua_State *L ) {
+	MOAI_LUA_SETUP ( MOAICCParticleSystem, "U" )
+	lua_pushnumber ( state, self->mRotPerSecondVariance );
```
rotatePerSecondVariance vs rotPerSecondVariance

kstenerud commented on src/moaicore/MOAICCParticleSystem.cpp:
```diff
@@ -0,0 +1,2389 @@
+	this->mLifespanTerm[0] = minLifespan;
+	
+	float maxLifespan = this->mLifespan + this->mLifespanVariance;
+	this->mLifespanTerm[1] = maxLifespan;
```
Why not just store these as mMinLifespan and mMaxLifespan?

kstenerud commented on src/moaicore/MOAICCParticleSystem.cpp:
```diff
@@ -0,0 +1,2389 @@
+MOAICCParticleSystem::~MOAICCParticleSystem () {
+	// clean up the allocated array
+	if (this->mParticles) {
+		free (this->mParticles);
```
These should not be allocated using the C allocator. Bad Things (tm) will happen.

kstenerud commented on src/moaicore/MOAICCParticleSystem.cpp:
```diff
@@ -0,0 +1,2389 @@
+	// clean up the allocated array
+	if (this->mParticles) {
+		free (this->mParticles);
+		this->mParticles = NULL;
```
This isn't necessary in a destructor.

kstenerud commented on src/moaicore/MOAICCParticleSystem.cpp:
```diff
@@ -0,0 +1,2389 @@
+	}
+	
+	if (this->mFlags & FLAGS_VISIBLE) {
+		for (int i = 0; i < (int) this->mParticleCount; ) {
```
This can be done more cleanly using C++ collections and iterators.

kstenerud commented on src/moaicore/MOAICCParticleSystem.cpp:
```diff
@@ -0,0 +1,2389 @@
+		
+		TiXmlElement* childElement = node->FirstChildElement ();
+		
+		fo
```
Use `new`, not `realloc`.

kstenerud commented on src/moaicore/MOAICCParticleSystem.h:
```diff
@@ -1,119 +1,141 @@
+	float		mDeckOffsetX;
+	float		mDeckOffsetY;
+	float		mDeckTileWidth;
+	float		mDeckTileHeight;
```
Indenting is off here.

kstenerud commented on src/moaicore/MOAICCParticleSystem.h:
```diff
@@ -188,34 +211,72 @@ class MOAIParticlePexPlugin :
-
+	static int		_startSystem						( lua_State* L );
+	static int		_stopSystem       					( lua_State* L );
+	static int      _surge								( lua_State* L );
```
Indenting

IkeBart replied:
It looks perfect in my copy of Xcode, but I realized that there is a mix of tabs and spaces.  The mix of tabs and spaces is probably causing the indentation problem on Github.

aab29 replied:
Could you make sure the spacing you commit matches the spacing the rest of Moai is using?

kstenerud commented on src/moaicore/MOAICCParticleSystem.h:
```diff
@@ -188,34 +211,72 @@ class MOAIParticlePexPlugin :
-	void			RegisterLuaFuncs			( MOAILuaState& state );
-};
+					MOAICCParticleSystem	();
+					~MOAICCParticleSystem	();
```
Indenting

kstenerud commented:
I'm a little worried about physically removing the old particle system. What is that going to do for our drift away from mainline moai?

aab29 commented:
Just a quck note:
This implementation was largely copied/stolen from Cocos2D, so a lot of it is the way it is because that's how Cocos2D did it.  :palm_tree: :gun:  I'm not saying we can't (or shouldn't) fix it up before merging it--I'm just explaining where some of the funky stuff came from, in case you were wondering.  :v: 

As for removing the old particle system, I'm pretty convinced it's the right thing to do.  :mortar_board:  The last time we made a change to Moai of more than a few lines was when we added our own text system alongside `MOAITextBox`.  :capital_abcd:  We left `MOAITextBox` in place to avoid creating conflicts with mainline Moai, but in practice, it's only complicated our own text system and created a ton of dead code.  :skull:  I'd like to avoid repeating that mistake.  :no_entry_sign: :repeat: 

Honestly, I'm doubtful that pulling in changes from mainline Moai is ever going to make sense for us.  :open_hands:  If others disagree, I'm happy to meet and discuss this point.  :round_pushpin:  But I'd rather talk about that IRL, since it's just as much about our long-term company goals as it is about architecture.  :tokyo_tower:

aab29 commented:
Looks like Isaac addressed all the comments but one on this branch.  :small_orange_diamond:  I just ran through all the games and made sure nothing broke.  He made another branch ( https://github.com/mindsnacks/moai-dev/tree/feature/ike-particles-std-vector ) where he used standard library vectors instead.  :ocean: I'm about to test all the games again on that implementation.  Assuming nothing breaks, we can pick whichever version we like better.  :nose:

IkeBart commented:
Before I made the changes to the Xcode projects, I was getting a linker error that was preventing me from testing.  More info about this in pull request #89.

IkeBart commented:
The test program for MOAICCParticleSystem works as expected under this branch.
